File: module-11-rl-agents/solutions/python/rl_training.py
"""Module 11: RL training loop."""
import os
import sys
import json

sys.path.insert(0, os.path.dirname(__file__))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-7-evals", "solutions", "python"))

from dotenv import load_dotenv
load_dotenv()

from environment import AgentEnvironment
from trajectories import collect_trajectory
from curriculum import Curriculum
from eval_types import EvalTask
from graders.code_grader import string_match_grader, file_exists_grader, composite_grader

easy_tasks = [
    EvalTask(
        id="easy-create-file", description="Create a simple file",
        prompt="Create a file called hello.txt containing 'Hello, World!'",
        setup=lambda d: None,
        grader=composite_grader([file_exists_grader("hello.txt"), string_match_grader("hello.txt", "Hello, World!")]),
    ),
    EvalTask(
        id="easy-edit-line", description="Change one line in a file",
        prompt="In greeting.ts, change the greeting from 'Hi' to 'Hello'.",
        setup=lambda d: open(os.path.join(d, "greeting.ts"), "w").write('export const greeting = "Hi";\n'),
        grader=string_match_grader("greeting.ts", '"Hello"'),
    ),
]

medium_tasks = [
    EvalTask(
        id="medium-add-function", description="Add a function to existing code",
        prompt="Add an 'isEven' function to math.ts that returns true if a number is even.",
        setup=lambda d: open(os.path.join(d, "math.ts"), "w").write("export function add(a: number, b: number) { return a + b; }\n"),
        grader=composite_grader([string_match_grader("math.ts", "isEven"), string_match_grader("math.ts", "% 2")]),
    ),
]

hard_tasks = [
    EvalTask(
        id="hard-fix-bug-and-test", description="Fix a bug and verify the fix",
        prompt="Fix the bug in sort.ts where the comparison is reversed, causing descending instead of ascending sort.",
        setup=lambda d: open(os.path.join(d, "sort.ts"), "w").write(
            "export function sortNumbers(arr: number[]): number[] {\n"
            "  return [...arr].sort((a, b) => b - a); // BUG: should be a - b\n}\n"
        ),
        grader=string_match_grader("sort.ts", "a - b"),
    ),
]

curriculum_tiers = [
    {"name": "Easy", "difficulty": "easy", "tasks": easy_tasks, "promotion_threshold": 0.8},
    {"name": "Medium", "difficulty": "medium", "tasks": medium_tasks, "promotion_threshold": 0.8},
    {"name": "Hard", "difficulty": "hard", "tasks": hard_tasks, "promotion_threshold": 0.8},
]


def main():
    max_episodes = int(sys.argv[1]) if len(sys.argv) > 1 else 6
    print("=== Module 11: RL Training Loop ===\n")
    print(f"Running {max_episodes} episodes with curriculum learning\n")

    curriculum = Curriculum(curriculum_tiers)
    env = AgentEnvironment()
    trajectories = []

    for ep in range(max_episodes):
        tier = curriculum.get_current_tier()
        task = curriculum.sample_task()

        print(f"\nEpisode {ep + 1}/{max_episodes} [{tier['name']}] Task: {task.id}")

        try:
            trajectory = collect_trajectory(
                env,
                {"files": {}, "prompt": task.prompt},
                task.grader,
            )
            trajectories.append(trajectory)

            outcome = curriculum.record_outcome(trajectory["success"])

            print(f"  Result: {'SUCCESS' if trajectory['success'] else 'FAIL'}")
            print(f"  Reward: {trajectory['reward']['total']:.3f} ({trajectory['reward']['breakdown']})")
            print(f"  Steps: {len(trajectory['steps'])}")

            if outcome["promoted"]:
                print(f"  Promoted to {outcome['new_tier']}!")
        except Exception as e:
            print(f"  Error: {e}")
            curriculum.record_outcome(False)

    state = curriculum.get_state()
    print("\n" + "=" * 60)
    print("RL TRAINING SUMMARY")
    print("=" * 60)
    print(f"Episodes: {state['episodes_completed']}")
    print(f"Successes: {state['total_successes']}")
    print(f"Current tier: {curriculum.get_current_tier()['name']}")
    print(f"Tier pass rates: {', '.join(f'{r * 100:.1f}%' for r in state['tier_pass_rates'])}")
    print(f"Curriculum complete: {curriculum.is_complete()}")

    output_path = os.path.join(os.getcwd(), "rl-trajectories.jsonl")
    with open(output_path, "w") as f:
        for t in trajectories:
            f.write(json.dumps({
                "episode_id": t["episode_id"],
                "prompt": t["prompt"],
                "steps": len(t["steps"]),
                "reward": t["reward"]["total"],
                "success": t["success"],
            }) + "\n")
    print(f"\nTrajectories saved to {output_path}")


if __name__ == "__main__":
    main()

File: module-11-rl-agents/solutions/rl-training.ts
/**
 * Module 11: RL training loop
 * Generate trajectories → compute rewards → report.
 *
 * Full policy updates require an RL framework (e.g., TRL, OpenRLHF).
 * This module demonstrates the data collection and reward computation pipeline.
 */
import "dotenv/config";
import * as fs from "fs/promises";
import * as path from "path";
import { AgentEnvironment } from "./environment.js";
import { collectTrajectory, type Trajectory } from "./trajectories.js";
import { Curriculum, type CurriculumTier } from "./curriculum.js";
import { graderAsReward, compositeRLVR } from "./rlvr.js";
import { stringMatchGrader, fileExistsGrader, compositeGrader } from "../../module-7-evals/solutions/graders/code-grader.js";
import type { EvalTask } from "../../module-7-evals/solutions/types.js";

// Define curriculum tiers
const easyTasks: EvalTask[] = [
  {
    id: "easy-create-file",
    description: "Create a simple file",
    prompt: "Create a file called hello.txt containing 'Hello, World!'",
    setup: async () => {},
    grader: compositeGrader([
      fileExistsGrader("hello.txt"),
      stringMatchGrader("hello.txt", "Hello, World!"),
    ]),
  },
  {
    id: "easy-edit-line",
    description: "Change one line in a file",
    prompt: "In greeting.ts, change the greeting from 'Hi' to 'Hello'.",
    setup: async (dir) => {
      await fs.writeFile(path.join(dir, "greeting.ts"), 'export const greeting = "Hi";\n');
    },
    grader: stringMatchGrader("greeting.ts", '"Hello"'),
  },
];

const mediumTasks: EvalTask[] = [
  {
    id: "medium-add-function",
    description: "Add a function to existing code",
    prompt: "Add an 'isEven' function to math.ts that returns true if a number is even.",
    setup: async (dir) => {
      await fs.writeFile(
        path.join(dir, "math.ts"),
        "export function add(a: number, b: number) { return a + b; }\n"
      );
    },
    grader: compositeGrader([
      stringMatchGrader("math.ts", "isEven"),
      stringMatchGrader("math.ts", "% 2"),
    ]),
  },
];

const hardTasks: EvalTask[] = [
  {
    id: "hard-fix-bug-and-test",
    description: "Fix a bug and verify the fix",
    prompt: "Fix the bug in sort.ts where the comparison is reversed, causing descending instead of ascending sort.",
    setup: async (dir) => {
      await fs.writeFile(
        path.join(dir, "sort.ts"),
        `export function sortNumbers(arr: number[]): number[] {
  return [...arr].sort((a, b) => b - a); // BUG: should be a - b
}
`
      );
    },
    grader: stringMatchGrader("sort.ts", "a - b"),
  },
];

const curriculumTiers: CurriculumTier[] = [
  { name: "Easy", difficulty: "easy", tasks: easyTasks, promotionThreshold: 0.8 },
  { name: "Medium", difficulty: "medium", tasks: mediumTasks, promotionThreshold: 0.8 },
  { name: "Hard", difficulty: "hard", tasks: hardTasks, promotionThreshold: 0.8 },
];

async function main() {
  const maxEpisodes = Number(process.argv[2]) || 6;
  console.log("=== Module 11: RL Training Loop ===\n");
  console.log(`Running ${maxEpisodes} episodes with curriculum learning\n`);

  const curriculum = new Curriculum(curriculumTiers);
  const env = new AgentEnvironment();
  const trajectories: Trajectory[] = [];

  for (let ep = 0; ep < maxEpisodes; ep++) {
    const tier = curriculum.getCurrentTier();
    const task = curriculum.sampleTask();

    console.log(`\nEpisode ${ep + 1}/${maxEpisodes} [${tier.name}] Task: ${task.id}`);

    try {
      const trajectory = await collectTrajectory(
        env,
        { files: {}, prompt: task.prompt },
        task.grader
      );

      // Note: we re-run setup inside collectTrajectory via the grader
      // In a real pipeline, environment.reset handles setup
      trajectories.push(trajectory);

      const { promoted, newTier } = curriculum.recordOutcome(trajectory.success);

      console.log(`  Result: ${trajectory.success ? "SUCCESS" : "FAIL"}`);
      console.log(`  Reward: ${trajectory.reward.total.toFixed(3)} (${trajectory.reward.breakdown})`);
      console.log(`  Steps: ${trajectory.steps.length}`);

      if (promoted) {
        console.log(`  🎉 Promoted to ${newTier}!`);
      }
    } catch (e) {
      console.error(`  Error: ${e instanceof Error ? e.message : String(e)}`);
      curriculum.recordOutcome(false);
    }
  }

  // Summary
  const state = curriculum.getState();
  console.log("\n" + "=".repeat(60));
  console.log("RL TRAINING SUMMARY");
  console.log("=".repeat(60));
  console.log(`Episodes: ${state.episodesCompleted}`);
  console.log(`Successes: ${state.totalSuccesses}`);
  console.log(`Current tier: ${curriculum.getCurrentTier().name}`);
  console.log(`Tier pass rates: ${state.tierPassRates.map((r) => (r * 100).toFixed(1) + "%").join(", ")}`);
  console.log(`Curriculum complete: ${curriculum.isComplete()}`);

  // Save trajectories
  const outputPath = path.join(process.cwd(), "rl-trajectories.jsonl");
  const lines = trajectories.map((t) => JSON.stringify({
    episodeId: t.episodeId,
    prompt: t.prompt,
    steps: t.steps.length,
    reward: t.reward.total,
    success: t.success,
  })).join("\n");
  await fs.writeFile(outputPath, lines, "utf-8");
  console.log(`\nTrajectories saved to ${outputPath}`);
}

main().catch(console.error);
